File: producers-consumers.py
'''
NAME: producers-consumers.py
AUTHOR: Andrew Meijer V00805554
In this implementation of Producer-consumer
I use the python Condition class found here:
https://docs.python.org/2/library/threading.html#condition-objects

'''

import threading
import asyncio
import random

# Global Variables
q = []
cv = asyncio.Condition()

# Pseudocode for Producer from Textbook:
# 1 event = waitForEvent()  - "select a random number"
# 2 mutex.wait()            - "enter critical section"
# 3 buffer.add(event)       - "produce the random number"
# 4 items.signal()          - "new item is on queue"
# 5 mutex.signal()          - "exit critical section"
def producer():
    nums = range(100)
    num = random.choice(nums)
    with cv:
        q.append(num)
        cv.notify()
    return

# Pseudocode for Consumer from Textbook:
# 1 items.wait()            -
# 2 mutex.wait()            -
# 3 event = buffer.get()    -
# 4 mutex.signal()          -
# 5 event.process()         -
def consumer():
    with cv:
        while not q:
            cv.wait()
        num = q.pop(0)
    print(num)
    return




#!/usr/bin/python3

import threading
import time

exitFlag = 0
q = []

class producer (threading.Thread):
   def __init__(self, threadID, name, counter):
      threading.Thread.__init__(self)
      self.threadID = threadID
      self.name = name
      self.counter = counter
   def run(self):
      print ("Starting " + self.name)
      print_time(self.name, self.counter, 5)
      print ("Exiting " + self.name)

class consumer (threading.Thread):
    def __init__(self, threadID, name, counter):
        threading.Thread.__init__(self)
        self.threadID = threadID
        self.name = name
        self.counter = counter
    def run(self):


# Create new threads
prod = myThread(1, "Thread-1", 1)
cons = myThread(2, "Thread-2", 2)

# Start new Threads
prod.start()
cons.start()
prod.join()
cons.join()
print ("Exiting Main Thread")

File: producers-consumers.c
//
// NAME:    producers-consumers.c
// AUTHOR:  Andrew Meijer V00805554
//  This program is based off the code from this URL:
//  https://www.geeksforgeeks.org/use-posix-semaphores-c/

//  I also found an implementation using processes rather than threads:
//  https://www.thecrazyprogrammer.com/2016/09/producer-consumer-problem-c.html


//  I use MinGW on Windows10 and compile with the following command:
//      gcc -pthread producers-consumers.c

#include <windows.h>
#include <stdio.h>
#include <pthread.h>
#include <semaphore.h>
#include <unistd.h>

sem_t mutex;
sem_t items;
//in this sample program, the buffer is only a single element.
int buffer = -1;
//~
short exitFlag = 0;
//startFlag keep program running until a production occurs
short startFlag = 0;

void* producer(void* arg)
{
    int n = 0;
    printf("\nStarting producer thread.\n");
    printf("Produce a number: ");
    scanf("%d",&n);

    startFlag = 1;

    sem_wait(&mutex);

    //critical section
    buffer = n;

    sem_post(&mutex);
    sem_post(&items);

    printf("Exiting producer thread.\n");
}

void* consumer(void* arg)
{
    int n = -1;
    while(!exitFlag){
        sem_wait(&items);
        sem_wait(&mutex);

        //critical section
        if(buffer != -1){
            n = buffer;
            buffer = -1;
            sem_post(&mutex);
            printf("Consumed %d\n", n);
        }else{
            sem_post(&mutex);
            //consumer waits for a production. 1fps-busywait
            sleep(1);
        }
    }
    printf("Exit flag is called. Consumption is over.\n");
}

int main()
{
    sem_init(&mutex, 0, 1);
    sem_init(&items, 0, 0);
    pthread_t prod,cons;
    pthread_create(&prod,NULL,producer,NULL);
    pthread_create(&cons,NULL,consumer,NULL);

    while(buffer != -1 || !startFlag){

    }

    exitFlag = 1;

    pthread_join(prod,NULL);
    pthread_join(cons,NULL);
    sem_destroy(&items);
    sem_destroy(&mutex);
    printf("Exiting Main Thread.\n");
}
